lib/dao.js

```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+
+  if (isPKMissing(Model, cb))
+    return cb.promise;
+
+
+  var strict = this.__strict;
+  var model = Model.modelName;
+  var hookState = {};
+
+  // Convert the data to be plain object so that update won't be confused
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
+
+  // Make sure id(s) cannot be changed
+  var idNames = Model.definition.idNames();
+  for (var i = 0, n = idNames.length; i < n; i++) {
+    var idName = idNames[i];
+    if (data[idName] !== undefined && !idEquals(data[idName], inst[idName])) {
+      var err = new Error('id property (' + idName + ') ' +
+              'cannot be updated from ' + inst[idName] + ' to ' + data[idName]);
+      err.statusCode = 400;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+  }
+  var context = {
+    Model: Model,
+    instance: inst,
+    currentInstance: inst,
+    hookState: hookState,
+    options: options
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    if (strict) {
+      self.strictChk(strict, data, inst, strictChkCB);
+    } else {
+      strictChkCB(null, data);
+    }
+
+    function strictChkCB(err, data) {
+      if (err) return cb(err);
+      // update instance's properties    
+      inst.setAttributes(data);
+
+      inst.isValid(function(valid) {
+        if (!valid) {
+          cb(new ValidationError(inst), inst);
+          return;
+        }
+
+        inst.trigger('save', function(saveDone) {
+          inst.trigger('update', function(done) {
+
+            var idNames = Model.definition.idNames();
+            // Convert idNames to an object since when we check ids the complexity would be O(1) in Object as opposed to an array which would be O(N)
+            // Please feel free to let me know if you dont like the idea
+            idNames = idNames.reduce(function(result, currentItem) {
+              result[currentItem] = currentItem;
+              return result;
+            }, {});
+
+            var propKeys = Object.keys(Model.definition.properties);
+            for (var i = 0; i < propKeys.length; i++) {
+              var p = propKeys[i];
+              if (idNames[p])
+                continue;
+              inst[p] = null;
+            }
+
+            copyData(data, inst);
+            var typedData = convertDataPropertiesByType(data);
+            context.data = typedData;
+
+            function replaceCallback(err, data) {
+              if (err) return cb(err);
+
+              var ctx = {
+                Model: Model,
+                instance: context.instance,
+                hookState: hookState,
+                options: options
+              };
+              Model.notifyObserversOf('loaded', ctx, function(err) {
+                if (err) return cb(err);
+
+                inst.__persisted = true;
+                inst.setAttributes(ctx.data);
```

Comment:
Please ensure there is a test covering this bit - that updates from the "loaded" hook are reflected in the instance returned to the callback.